File: blas/hsum_float32_arm64.go
package blas

import (
	"unsafe"
)

//go:noescape

func _hsum_float32(input, result unsafe.Pointer, size uint64)
func HsumFloat32(input []float32) float32 {
	var sum float32
	_hsum_float32(unsafe.Pointer(&input[0]), unsafe.Pointer(&sum), uint64(len(input)))
	return sum
}

File: blas/cpp/hsum_float32_arm64.cpp
#include <arm_neon.h>

void hsum_float32(float *input, float *result, uint64_t size) {
    *result = 0.0;
    int chunks = size / 4;

    for (int i = 0; i < chunks; i++) {
        *result += vaddvq_f32(((float32x4_t *)input)[i]);
    }
    // Leftovers
    for (int i = chunks * 4; i < size; i++) {
        *result +=  input[i];
    }
}
